Parser crash (`AssertionError`) when line comment hides then in for-generator inside new argument

## Compiler version

- 3.6.4 (latest stable release)
- 3.8.3-RC1-SNAPSHOT (nightly)

## Minimized code

```scala
//> using scala 3.6.4

val y = new c(for baz <- if baz // x then y else new A {
  val z = w } do 1)
```

The line comment `// x then y else new A {` causes the parser to see `if baz` without a `then` clause. During error recovery, the parser constructs an AST where the for-comprehension's span extends one byte beyond its parent `new c(...)` span.

## Output

```
-- [E040] Syntax Error: crash.scala:2:2 --
2 |  val z = w } do 1)
  |  ^^^
  |  'then' expected, but 'val' found

  exception occurred while parser crash.scala
  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/scala/scala3/issues/new/choose

     while compiling: crash.scala
        during phase: parser
    compiler version: version 3.6.4

Exception in thread "main" java.lang.AssertionError: assertion failed: position error, parent span does not contain child span
parent      = new c(for baz <- if baz then ??? do ???) # -1,
parent span = <8..76>,
child       = for baz <- if baz then ??? do ??? # -1,
child span  = [14..77]
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.ast.Positioned.check$1(Positioned.scala:177)
	at dotty.tools.dotc.ast.Positioned.check$1$$anonfun$3(Positioned.scala:207)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.ast.Positioned.check$1(Positioned.scala:207)
	at dotty.tools.dotc.ast.Positioned.checkPos(Positioned.scala:228)
	at dotty.tools.dotc.ast.Positioned.check$1(Positioned.scala:202)
	at dotty.tools.dotc.ast.Positioned.checkPos(Positioned.scala:228)
	at dotty.tools.dotc.ast.Positioned.check$1(Positioned.scala:202)
	at dotty.tools.dotc.ast.Positioned.check$1$$anonfun$3(Positioned.scala:207)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.ast.Positioned.check$1(Positioned.scala:207)
	at dotty.tools.dotc.ast.Positioned.checkPos(Positioned.scala:228)
	at dotty.tools.dotc.parsing.Parser.parse$$anonfun$1(ParserPhase.scala:39)
	at dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:510)
	at dotty.tools.dotc.parsing.Parser.parse(ParserPhase.scala:40)
	...
```

## Expectation

The compiler should parse erroneous programs without crashing. Syntax errors should be reported gracefully, and the AST position invariant (parent span contains all child spans) should be maintained even during error recovery.

Instead, the parser crashes with an `AssertionError` when checking position consistency of the parsed AST.

### Analysis

The crash occurs in `Positioned.check` (`Positioned.scala:177`) during the parser phase:

1. The line comment `//` on line 1 hides `then y else new A {`, so the parser sees `for baz <- if baz` — an `if` without a `then` clause.
2. On line 2, the parser encounters `val z = w } do 1)` and triggers error recovery (expecting `then`, got `val`).
3. During error recovery, the parser reconstructs the AST but assigns the for-comprehension a span of `[14..77]` while its parent `new c(...)` gets a span of `<8..76>` — the child's end position (77) exceeds the parent's end position (76) by 1 byte.
4. The `checkPos` validation in `Positioned.scala` detects this invariant violation and throws `AssertionError`.

This is a parser-phase bug - the crash happens before type checking, purely during syntax error recovery. The root cause is likely in how the error recovery mechanism computes source positions when an `if`-expression inside a `for`-generator is missing its `then` clause due to a line comment.

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -2929,7 +2929,9 @@ object Parsers {
      */
     def newExpr(): Tree =
       val start = in.skipToken()
-      def reposition(t: Tree) = t.withSpan(Span(start, in.lastOffset))
+      def reposition(t: Tree) =
+        val end = if t.span.exists then in.lastOffset max t.span.end else in.lastOffset
+        t.withSpan(Span(start, end))
       possibleTemplateStart()
       val parents =
         if in.isNestedStart then Nil
@@ -2943,11 +2945,12 @@ object Parsers {
           else parent
       case parents =>
         // With brace syntax, the last token consumed by a parser is }, but with indent syntax,
-        // the last token consumed by a parser is OUTDENT, which causes mismatching spans, so don't reposition.
+        // the last token consumed by a parser is OUTDENT, which causes mismatching spans, so don't update end.
         val indented = in.token == INDENT
         val body =
-          val bo = templateBodyOpt(emptyConstructor, parents, derived = Nil)
-          if !indented then reposition(bo) else bo
+          val tbo = templateBodyOpt(emptyConstructor, parents, derived = Nil)
+          if indented then tbo.withSpan(tbo.span.withStart(start))
+          else reposition(tbo)
         New(body)
     end newExpr
 

--- a/tests/neg-custom-args/captures/mut-iterator4.check
+++ b/tests/neg-custom-args/captures/mut-iterator4.check
@@ -3,9 +3,9 @@
   |                          ^^^^^^^^^^^^^
   |                          Read-only method map accesses exclusive capability `Iterator.this`;
   |                          method map should be declared an update method to allow this.
--- [E007] Type Mismatch Error: tests/neg-custom-args/captures/mut-iterator4.scala:21:81 --------------------------------
+-- [E007] Type Mismatch Error: tests/neg-custom-args/captures/mut-iterator4.scala:21:77 --------------------------------
 21 |def mappedIterator[T, U](it: Iterator[T]^, f: T => U): Iterator[U]^{it, f} = new Iterator: // error
-   |                                                                                 ^
+   |                                                                             ^
    |Found:    Iterator[U^'s1]^{it.rd, f, it, any}
    |Required: Iterator[U]^{it, f}
    |

--- a/tests/neg/i25248.scala
+++ b/tests/neg/i25248.scala
@@ -1,0 +1,2 @@
+val y = new c(for baz <- if baz // x then y else new A { // error
+  val z = w } do 27) 42 // error

--- a/tests/semanticdb/expect/StructuralTypes.expect.scala
+++ b/tests/semanticdb/expect/StructuralTypes.expect.scala
@@ -16,7 +16,7 @@ object StructuralTypes/*<-example::StructuralTypes.*/:
 
   val V/*<-example::StructuralTypes.V.*/: Object/*->java::lang::Object#*/ {
     def scalameta/*<-local4*/: String/*->scala::Predef.String#*/
-  } = new:
-    /*<-local6*/def scalameta/*<-local5*/ = "4.0"
+  } = /*<-local6*/new:
+    def scalameta/*<-local5*/ = "4.0"
   V/*->example::StructuralTypes.V.*/.scalameta/*->scala::reflect::Selectable#selectDynamic().*/
 end StructuralTypes/*->example::StructuralTypes.*/

--- a/tests/semanticdb/metac.expect
+++ b/tests/semanticdb/metac.expect
@@ -3810,7 +3810,7 @@ Occurrences:
 [16:9..16:15): Object -> java/lang/Object#
 [17:8..17:17): scalameta <- local4
 [17:19..17:25): String -> scala/Predef.String#
-[19:4..19:4): <- local6
+[18:6..18:6): <- local6
 [19:8..19:17): scalameta <- local5
 [20:2..20:3): V -> example/StructuralTypes.V.
 [20:4..20:13): scalameta -> scala/reflect/Selectable#selectDynamic().

